Userland/Libraries/LibJS/Runtime/VM.cpp

```diff
@@ -856,4 +894,111 @@ ThrowCompletionOr<NonnullRefPtr<Module>> VM::resolve_imported_module(ScriptOrMod
+    auto fulfilled_closure = [&referencing_script_or_module, &specifier, &promise_capability](VM& vm, GlobalObject& global_object) -> ThrowCompletionOr<Value> {
+        auto result = vm.argument(0);
+        // a. Assert: result is undefined.
+        VERIFY(result.is_undefined());
+        // b. Let moduleRecord be ! HostResolveImportedModule(referencingScriptOrModule, specifier).
+        auto module_record = MUST(vm.host_resolve_imported_module(referencing_script_or_module, specifier));
+
+        // c. Assert: Evaluate has already been invoked on moduleRecord and successfully completed.
+        // Note: If HostResolveImportedModule returns a module evaluate will have been called on it.
+
+        // d. Let namespace be GetModuleNamespace(moduleRecord).
+        auto namespace_ = module_record->get_module_namespace(vm);
+
+        VERIFY(!vm.exception());
+        // e. If namespace is an abrupt completion, then
+        if (namespace_.is_throw_completion()) {
+            // i. Perform ! Call(promiseCapability.[[Reject]], undefined, « namespace.[[Value]] »).
+            MUST(JS::call(global_object, promise_capability.reject, js_undefined(), *namespace_.throw_completion().value()));
+        }
+        // f. Else,
+        else {
+            // i. Perform ! Call(promiseCapability.[[Resolve]], undefined, « namespace.[[Value]] »).
+            MUST(JS::call(global_object, promise_capability.resolve, js_undefined(), namespace_.release_value()));
+        }
+        // g. Return undefined.
+        return js_undefined();
+    };
+
+    // 2. Let onFulfilled be ! CreateBuiltinFunction(fulfilledClosure, 0, "", « »).
+    auto* on_fulfilled = NativeFunction::create(current_realm()->global_object(), "", move(fulfilled_closure));
+
+    // 3. Let rejectedClosure be a new Abstract Closure with parameters (error) that captures promiseCapability and performs the following steps when called:
+    auto rejected_closure = [&promise_capability](VM& vm, GlobalObject& global_object) -> ThrowCompletionOr<Value> {
+        auto error = vm.argument(0);
+        // a. Perform ! Call(promiseCapability.[[Reject]], undefined, « error »).
+        MUST(JS::call(global_object, promise_capability.reject, js_undefined(), error));
+        // b. Return undefined.
+        return js_undefined();
+    };
+
+    // 4. Let onRejected be ! CreateBuiltinFunction(rejectedClosure, 0, "", « »).
+    auto* on_rejected = NativeFunction::create(current_realm()->global_object(), "", move(rejected_closure));
+
+    // 5. Perform ! PerformPromiseThen(innerPromise, onFulfilled, onRejected).
+    inner_promise->perform_then(on_fulfilled, on_rejected, {});
+
+    VERIFY(!exception());
+
+    // Note: We handle these immediately as we don't have asynchronous processing yet.
+    run_queued_promise_jobs();
```

Comment:
Is this actually needed? `Interpreter::run()` / the host should do it for us.

Reply:
I thought we needed this one but that was because I used references in the lambdas :facepalm: